builder/processor/src/main/java/io/helidon/builder/processor/FactoryMethods.java

```diff
@@ -0,0 +1,329 @@
+ * @param createFromConfig
+ * @param builder
+ */
+record FactoryMethods(Optional<FactoryMethod> createTargetType,
+                      Optional<FactoryMethod> createFromConfig,
+                      Optional<FactoryMethod> builder) {
+    static FactoryMethods create(ProcessingContext processingContext,
+                                 TypeInfo blueprint,
+                                 TypeHandler typeHandler) {
+
+        Optional<FactoryMethod> targetFactory = targetTypeMethod(processingContext, blueprint, typeHandler);
+        Set<TypeName> configObjectCandidates = new LinkedHashSet<>();
+        if (targetFactory.isPresent()) {
+            configObjectCandidates.add(targetFactory.get().argumentType());
+        }
+        configObjectCandidates.add(typeHandler.actualType());
+        configObjectCandidates.add(typeHandler.declaredType());
+
+        // the candidate from factory method is first, as it is more significant
+        Optional<FactoryMethod> configFactory = createFromConfigMethod(processingContext,
+                                                                       blueprint,
+                                                                       typeHandler,
+                                                                       configObjectCandidates);
+        configObjectCandidates = new LinkedHashSet<>();
+        if (targetFactory.isPresent()) {
+            configObjectCandidates.add(targetFactory.get().argumentType());
+        }
+        if (configFactory.isPresent()) {
+            configObjectCandidates.add(configFactory.get().factoryMethodReturnType());
+        }
+
+        return new FactoryMethods(targetFactory,
+                                  configFactory,
+                                  builder(processingContext, blueprint, typeHandler, configObjectCandidates));
+    }
+
+    private static Optional<FactoryMethod> builder(ProcessingContext processingContext,
+                                                   TypeInfo blueprint,
+                                                   TypeHandler typeHandler,
+                                                   Set<TypeName> builderCandidates) {
+        if (typeHandler.actualType().equals(OBJECT_TYPE)) {
+            return Optional.empty();
+        }
+        builderCandidates.add(typeHandler.actualType());
+        FactoryMethod found = null;
+        FactoryMethod secondary = null;
+        for (TypeName builderCandidate : builderCandidates) {
+            if (typeHandler.actualType().primitive()) {
+                // primitive methods do not have builders
+                found = null;
+                continue;
+            }
+            TypeInfo typeInfo = processingContext.toTypeInfo(builderCandidate.genericTypeName()).orElse(null);
+            if (typeInfo == null) {
+                if (secondary == null) {
+                    // this may be part of annotation processing where type info is not available
```

Comment:
Same thing as above - we may not have full information about a type that is generated as part of this annotation processing round (and there may be cyclic dependencies between types to do this in multiple processing rounds), so sometimes we just make an assumption you do it right.
The worst case scenario is that this will fail to compile, and you will get a proper compiler error explaining why it failed and which generated type caused the failure